### DISCUSSION TITLE
Save As PDF

### DISCUSSION THREAD

**Discussion by NCHernyshov** [ShapeCrawler/ShapeCrawler#602] [Q&A]
Hello. Can you tell me how I can save the presentation in PDF format?

**Comment by AdamShakhabov** [comment]
Hi.

I think you should use a commercial library. ShapeCrawler does not have such feature. Also, I don't believe there is an open-source C# library with this feature. If you find one, please let me know🙂

**Reply by NickKellett** [reply]
I run ShapeCrawler (awesome library!) and save my powerpoint. Then I use a local installation of [LibreOffice ](https://www.libreoffice.org/) to save as PDF. In C# you can invoke a command line process to do this:

```
var exe = FindLibreOfficeExe(libreOfficeExeRootFilePath);
var psi = new ProcessStartInfo
{
    FileName = exe,
    // Common flags to keep it quiet and reliable:
    Arguments = $"--headless --nologo --nolockcheck --norestore --convert-to pdf \"{inputFull}\" --outdir \"{tempOutDir}\"",
    RedirectStandardOutput = true,
    RedirectStandardError = true,
    UseShellExecute = false,
    CreateNoWindow = true,
    WorkingDirectory = inputDir
};

using var proc = Process.Start(psi) ?? throw new InvalidOperationException("Failed to start LibreOffice process.");
string stdout = proc.StandardOutput.ReadToEnd();
string stderr = proc.StandardError.ReadToEnd();
proc.WaitForExit();

if (proc.ExitCode != 0)
    throw new Exception($"LibreOffice conversion failed (exit {proc.ExitCode}).\nSTDOUT:\n{stdout}\nSTDERR:\n{stderr}");
```

Of course you could also do scripting.

**Reply by AdamShakhabov** [reply]
Thanks a lot for your feedback, @NickKellett ☺️ We also used LibreOffice in this way on one of my main jobs to create slide images.

**Reply by NickKellett** [reply]
I saw that you are working on that - I'm looking to save as images too, so let me know if I can do any testing for you when's that ready, or help in the meantime.

**Reply by AdamShakhabov** [reply]
Right. I believe it is time to implement slide image generation, and yes, we will need a lot of testing on real use cases.
